### ISSUE TITLE
used-before-assignment false positive

### ISSUE THREAD

**Issue by nikita-leontiev** [pylint-dev/pylint#9126]
### Bug description

```python
Code below leads to used-before-assignment false positive:

def function(*args):
    print(args)

def wrapper(arg):
    return lambda *args: function(arg, *args)

if __name__ == "__main__":
    args = None
    func = wrapper(1)
    func(2)
```
W0621 warning here is expected.
```


### Configuration

```ini
Without pylintrc
```

### Command used

```shell
pylint -sn --disable=C0103,C0114,C0116 test.py
```

### Pylint output

```python
************* Module test
test.py:3:0: W0621: Redefining name 'args' from outer scope (line 10) (redefined-outer-name)
test.py:7:40: E0601: Using variable 'args' before assignment (used-before-assignment)
```

### Expected behavior

Only one W0621 warning.

### Pylint version

```shell
pylint 3.0.1
astroid 3.0.0
Python 3.8.10 (default, May 26 2023, 14:05:08)
[GCC 9.4.0]
```

### OS / Environment

Ubuntu 20.04

**Comment by pianofab** [comment]
I am seeing a similar “false positive” for a typical use case (?). Let’s say I have the traditional “if __name__ == ‘__main__’:” at the bottom of my main module, and the first line inside the if statement is: “logger = logging.getLogger(__file__)”. I see a false positive when I use logger in one place; this could be true if my module is imported but the outside code does not have a logger defined, but do we care?
